### PR TITLE
[feat] Prevent double spending: UTXOs created and spent in the same block

### PR DESCRIPTION
**Pull request by TAdev0**
- [x] issue #164 
- [x] follows contribution [guide](https://github.com/keep-starknet-strange/raito/blob/main/CONTRIBUTING.md)
- [ ] code change includes tests

### CODE REVIEW

File: src/types/state.cairo [hunk 1]
```diff
@@ -0,0 +1,10 @@
+//! State is a top level struct containing the chain state and the utxo set
+
+use super::chain_state::ChainState;
+use super::utxo_set::UtxoSet;
+
+#[derive(Default, Destruct)]
+pub struct State {
+    chain_state: ChainState,
+    utxo_set: UtxoSet
```
Comment (m-kus) [inline src/types/state.cairo:9]: Here we only need UtreexoState (roots + number of leaves)
UtxoSet also contains local cache (which is not part of input/output) + probably some additional validation context

**Comment by TAdev0** [comment]
@m-kus  @maciejka  before i implement the double spending prevention functionality, could you tell me using `State` this way is ok for you?

File: packages/consensus/src/types/state.cairo [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+pub struct State {
+    pub chain_state: ChainState,
+    pub utreexo_state: UtreexoState
+}
+
+/// Full block validator (w/o bitcoin script checks and utxo inclusion verification for now).
+#[generate_trait]
+pub impl BlockValidatorImpl of BlockValidator {
+    fn validate_and_apply(self: State, block: Block) -> Result<State, ByteArray> {
+        let utxo_set = UtxoSet {
```
Comment (TAdev0) [inline packages/consensus/src/types/state.cairo:23]: for now, just instantiated the `utxo_set` using the utreexo from the global state, waiting for your review to finish implementing the feature of my issue

File: packages/consensus/src/types/state.cairo [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+        validate_timestamp(self.chain_state.prev_timestamps, block.header.time)?;
+        let prev_block_time = *self.chain_state.prev_timestamps[self.chain_state.prev_timestamps.len() - 1];
+        let prev_timestamps = next_prev_timestamps(self.chain_state.prev_timestamps, block.header.time);
+
+        let txid_root = match block.data {
+            TransactionData::MerkleRoot(root) => root,
+            TransactionData::Transactions(txs) => {
+                let (total_fees, txid_root, wtxid_root) = compute_and_validate_tx_data(
+                    txs, block_height, block.header.time, ref utxo_set
```
Comment (TAdev0) [inline packages/consensus/src/types/state.cairo:38]: passed by reference `utxo_set`  to `compute_and_validate_tx_data`

File: packages/consensus/src/types/state.cairo [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+pub struct State {
+    pub chain_state: ChainState,
+    pub utreexo_state: UtreexoState
+}
+
+/// Full block validator (w/o bitcoin script checks and utxo inclusion verification for now).
+#[generate_trait]
+pub impl BlockValidatorImpl of BlockValidator {
+    fn validate_and_apply(self: State, block: Block) -> Result<State, ByteArray> {
+        let mut utxo_set = UtxoSet {
```
Comment (TAdev0) [inline packages/consensus/src/types/state.cairo:23]: instantiate `utxo_set` using `self. utreexo_state` and a new default cache

Comment (TAdev0) [reply packages/consensus/src/types/state.cairo:23]: I can stick to the scope of the issue and instantiate a dummy utreexo but i felt i would be better to modify the whole flow already and use the global state

**Comment by TAdev0** [comment]
I curiously have the 2 first integration tests failing with
```
        tests/data/full_169.json — Panicked with 'Negative fee (output 5000000000 > input 0)'
        tests/data/full_757738.json — Panicked with 'Negative fee (output 625107042 > input 0)'
 ```
 
 all other are passing. Trying to understand as my code isnt supposd to impact fee

File: packages/client/src/test.cairo [hunk 5]
```diff
@@ -22,13 +23,15 @@ fn test(mut arguments: Span<felt252>) {
     let Args { mut chain_state, blocks, expected_chain_state } = Serde::deserialize(ref arguments)
         .expect('Failed to deserialize');
 
+    let mut state: State = State { chain_state: chain_state, utreexo_state: Default::default(), };
```
Comment (TAdev0) [inline packages/client/src/test.cairo:26]: temporary solution while python scripts are not updated.
Will only work if we test one isolated block, or a batch of block starting from genesis block as instantiate a default `utreexo_state` in the `state`

**Comment by TAdev0** [comment]
@maciejka  @m-kus  ready for final review!

File: packages/consensus/src/types/state.cairo [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            State {
+                chain_state: ChainState {
+                    block_height,
+                    total_work,
+                    best_block_hash,
+                    current_target,
+                    epoch_start_time,
+                    prev_timestamps,
+                },
+                utreexo_state: utxo_set.utreexo_state
```
Comment (TAdev0) [inline packages/consensus/src/types/state.cairo:75]: returns the full state, contaning the new chainstate + `utxo_set.utreexo_state`

File: packages/consensus/src/validation/block.cairo [hunk 7]
```diff
@@ -47,6 +50,50 @@ pub fn compute_and_validate_tx_data(
 
         let txid = double_sha256_byte_array(tx_bytes_legacy);
 
+        // Removes the outpoint hash of a transaction input if it was in the cache.
```
Comment (m-kus) [inline packages/consensus/src/validation/block.cairo:53]: It’s better to put this logic inside validate_transactions where we already have loops over inputs and outputs (to minimize number of loops).

File: packages/consensus/src/validation/block.cairo [hunk 8]
```diff
@@ -47,6 +50,50 @@ pub fn compute_and_validate_tx_data(
...
+        let inputs = *tx.inputs;
+        let mut j = 0;
+        while j != inputs.len() {
+            let outpoint = (*inputs[j]).previous_output;
+            let outpoint_hash = PoseidonTrait::new().update_with(outpoint).finalize();
```
Comment (m-kus) [inline packages/consensus/src/validation/block.cairo:58]: The hashing and updating the cache should be encapsulated in utxoset

File: packages/consensus/src/validation/block.cairo (hunk 7 above)
Comment (TAdev0) [reply packages/consensus/src/validation/block.cairo:53]: if i do that, then coinbase is excluded, i should handle it separately ?

Comment (TAdev0) [reply packages/consensus/src/validation/block.cairo:53]: ah you mean only inputs loop right?

File: packages/consensus/src/validation/block.cairo (hunk 8 above)
Comment (TAdev0) [reply packages/consensus/src/validation/block.cairo:58]: what do you mean?

File: packages/consensus/src/validation/block.cairo (hunk 7 above)
Comment (TAdev0) [reply packages/consensus/src/validation/block.cairo:53]: well i suppose its not needed as coinbase tx is treated separately, and will never be cached as it cannot be spend in the same block and there will be a check for that

**Review by m-kus — APPROVED** [review]

**Pull request merged**